Autostartup + minimized

Noxis-source opened the pull request:
Added autostartup and minimized options

Noxis-source commented:
https://github.com/jliddev/WowUp/issues/214

Noxis-source commented:
I approached this task in two different ways:
- [app.setLoginItemSettings](https://www.electronjs.org/docs/api/app#appsetloginitemsettingssettings-macos-windows)
- [node-auto-launch](https://github.com/Teamwork/node-auto-launch)

Here is a short comparison table demonstrating testing results:

Approach | Windows | MacOs | Linux |
---|---|---|---|
№1 | +/- | +/+ | -/- |
№2 | +/- | ?/? | ?/? |

So, the first way does not support Linux at all, while the second claims the capability to do what is needed. I tested №1 on all platforms but did not have a chance to do that for №2. It works on Windows as expected. I also implemented a w/a for the minimized start feature from here: https://github.com/electron/electron/issues/6893#issuecomment-300463668.

Noxis-source commented on wowup-electron/main.ts:
```diff
@@ -230,7 +230,9 @@ function createWindow(): BrowserWindow {
   win.webContents.userAgent = USER_AGENT;
 
   win.once("ready-to-show", () => {
-    win.show();
```
A w/a for minimized startup on Windows

Noxis-source commented on wowup-electron/package.json:
```diff
@@ -114,6 +114,7 @@
     "@types/lodash": "4.14.162",
     "adm-zip": "0.4.16",
     "async": "3.2.0",
```
Reference for auto-launch

Noxis-source commented on wowup-electron/src/app/pages/options/options.component.html:
```diff
@@ -60,6 +60,24 @@
             category="Options" action="UseHardwareAcceleration" [label]="useHardwareAcceleration">
           </mat-slide-toggle>
         </div>
```
New switches in the Options view
Ideally, I would make the second one disabled depending on the state of the first one. However, I just did not find a way to do that besides using this thing: https://angular.io/api/forms/NgModel. Yet I did not see anyting like this in the markup, so I decided to leave it as is - this does not really hurt anything. Maybe switches' text can be modified to be more clear in this regard.

Noxis-source commented on wowup-electron/src/app/pages/options/options.component.ts:
```diff
@@ -34,6 +34,8 @@ export class OptionsComponent implements OnInit, OnChanges {
   public collapseToTray = false;
   public telemetryEnabled = false;
   public useHardwareAcceleration = true;
```
New bool fields

Noxis-source commented on wowup-electron/src/app/pages/options/options.component.ts:
```diff
@@ -130,6 +132,14 @@ export class OptionsComponent implements OnInit, OnChanges {
     });
   };
 
```
Change handlers for switches

Noxis-source commented on wowup-electron/src/app/pages/options/options.component.ts:
```diff
@@ -143,6 +153,8 @@ export class OptionsComponent implements OnInit, OnChanges {
       this.telemetryEnabled = this._analyticsService.telemetryEnabled;
       this.collapseToTray = this.wowupService.collapseToTray;
       this.useHardwareAcceleration = this.wowupService.useHardwareAcceleration;
```
Set fields from preferences

Noxis-source commented on wowup-electron/src/app/services/wowup/wowup.service.ts:
```diff
@@ -27,9 +27,12 @@ import {
   LAST_SELECTED_WOW_CLIENT_TYPE_PREFERENCE_KEY,
   WOWUP_RELEASE_CHANNEL_PREFERENCE_KEY,
   USE_HARDWARE_ACCELERATION_PREFERENCE_KEY,
```
New preference keys

Noxis-source commented on wowup-electron/src/app/services/wowup/wowup.service.ts:
```diff
@@ -27,9 +27,12 @@ import {
+  START_MINIMIZED_PREFERENCE_KEY
 } from "common/constants";
 
 const LATEST_VERSION_CACHE_KEY = "latest-version-response";
```
auto-launch reference

Noxis-source commented on wowup-electron/src/app/services/wowup/wowup.service.ts:
```diff
@@ -103,6 +106,36 @@ export class WowUpService {
     this._preferenceChangeSrc.next({ key, value: value.toString() });
   }
 
```
New properties synced to preferences

Noxis-source commented on wowup-electron/src/app/services/wowup/wowup.service.ts:
```diff
@@ -311,4 +344,16 @@ export class WowUpService {
       this.applicationDownloadsFolderPath
     );
   }
```
Set autostartup options there

Noxis-source commented on wowup-electron/src/assets/i18n/en.json:
```diff
@@ -133,11 +133,15 @@
-        "USE_HARDWARE_ACCELERATION_DESCRIPTION": "Disabling hardware acceleration might solve FPS issues and fix other rendering issues in this app. Changing this setting requires a restart.",
-        "USE_HARDWARE_ACCELERATION_DISABLE_CONFIRMATION_DESCRIPTION": "Disabling hardware acceleration requires the application to restart.",
-        "USE_HARDWARE_ACCELERATION_ENABLE_CONFIRMATION_DESCRIPTION": "Enabling hardware acceleration requires the application to restart.",
-        "USE_HARDWARE_ACCELERATION_LABEL": "Enable Hardware Acceleration"
```
Labels and descriptions for new options

Noxis-source commented on wowup-electron/src/assets/i18n/ru.json:
```diff
@@ -133,11 +133,15 @@
-        "USE_HARDWARE_ACCELERATION_DESCRIPTION": "Отключение аппаратного ускорения может решить проблемы с количеством кадров в секунду и исправить другие неполадки с приложением. Изменение этой настройки требует перезапуск приложения.",
-        "USE_HARDWARE_ACCELERATION_DISABLE_CONFIRMATION_DESCRIPTION": "Для отключения аппаратного ускорения нужен перезапуск приложения.",
-        "USE_HARDWARE_ACCELERATION_ENABLE_CONFIRMATION_DESCRIPTION": "Для включения аппаратного ускорения нужен перезапуск приложения.",
-        "USE_HARDWARE_ACCELERATION_LABEL": "Включить аппаратное ускорение"
```
Russian localization

Noxis-source commented on wowup-electron/src/common/constants.ts:
```diff
@@ -25,7 +25,7 @@ export const LAST_SELECTED_WOW_CLIENT_TYPE_PREFERENCE_KEY =
 export const USE_HARDWARE_ACCELERATION_PREFERENCE_KEY =
   "use_hardware_acceleration";
-
-// ERRORS
```
Constants for new options

Noxis-source reviewed (commented):
Some notes

Noxis-source commented on wowup-electron/main.ts:
```diff
@@ -230,7 +230,9 @@ function createWindow(): BrowserWindow {
-    win.show();
+    var startMinimized = (process.argv || []).indexOf('--hidden') !== -1;
+    if (!startMinimized)
+      win.show();
```
A w/a for minimized startup on Windows

Noxis-source commented on wowup-electron/package.json:
```diff
@@ -114,6 +114,7 @@
     "@types/lodash": "4.14.162",
     "adm-zip": "0.4.16",
     "async": "3.2.0",
+    "auto-launch": "5.0.5",
```
Reference for auto-launch

Noxis-source commented on wowup-electron/src/app/pages/options/options.component.html:
```diff
@@ -60,6 +60,24 @@
+          <mat-slide-toggle [(checked)]="startMinimized" (change)="onStartMinimizedChange($event)" appUserActionTracker
+            category="Options" action="StartMinimized" [label]="startMinimized">
+          </mat-slide-toggle>
+        </div>
```
New switches in the Options view
Ideally, I would make the second one disabled depending on the state of the first one. However, I just did not find a way to do that besides using this thing: https://angular.io/api/forms/NgModel. Yet I did not see anyting like this in the markup, so I decided to leave it as is - this does not really hurt anything. Maybe switches' text can be modified to be more clear in this regard.

Noxis-source replied:
@jliddev Can you please also look into this? Is there anything I missed about how to make this work?

Noxis-source replied:
Nvm :)

Noxis-source commented on wowup-electron/src/app/pages/options/options.component.ts:
```diff
@@ -34,6 +34,8 @@ export class OptionsComponent implements OnInit, OnChanges {
   public telemetryEnabled = false;
   public useHardwareAcceleration = true;
+  public startWithSystem = false;
+  public startMinimized = false;
```
New bool fields

Noxis-source commented on wowup-electron/src/app/pages/options/options.component.ts:
```diff
@@ -130,6 +132,14 @@ export class OptionsComponent implements OnInit, OnChanges {
+  onStartMinimizedChange = (evt: MatSlideToggleChange) => {
+    this.wowupService.startMinimized = evt.checked;
+  };
+
```
Change handlers for switches

Noxis-source commented on wowup-electron/src/app/pages/options/options.component.ts:
```diff
@@ -143,6 +153,8 @@ export class OptionsComponent implements OnInit, OnChanges {
       this.collapseToTray = this.wowupService.collapseToTray;
       this.useHardwareAcceleration = this.wowupService.useHardwareAcceleration;
+      this.startWithSystem = this.wowupService.startWithSystem;
+      this.startMinimized = this.wowupService.startMinimized;
```
Set fields from preferences

Noxis-source commented on wowup-electron/src/app/services/wowup/wowup.service.ts:
```diff
@@ -27,9 +27,12 @@ import {
   WOWUP_RELEASE_CHANNEL_PREFERENCE_KEY,
   USE_HARDWARE_ACCELERATION_PREFERENCE_KEY,
+  START_WITH_SYSTEM_PREFERENCE_KEY,
+  START_MINIMIZED_PREFERENCE_KEY
```
New preference keys

Noxis-source commented on wowup-electron/src/app/services/wowup/wowup.service.ts:
```diff
@@ -27,9 +27,12 @@ import {
 } from "common/constants";
 
 const LATEST_VERSION_CACHE_KEY = "latest-version-response";
+var autoLaunch = require('auto-launch');
```
auto-launch reference

Noxis-source commented on wowup-electron/src/app/services/wowup/wowup.service.ts:
```diff
@@ -103,6 +106,36 @@ export class WowUpService {
+
+    this.setAutoStartup();
+  }
+
```
New properties synced to preferences

Noxis-source commented on wowup-electron/src/app/services/wowup/wowup.service.ts:
```diff
@@ -311,4 +344,16 @@ export class WowUpService {
+      autoLauncher.enable();
+    else
+      autoLauncher.disable();
+  }
```
Set autostartup options there

Noxis-source commented on wowup-electron/src/assets/i18n/en.json:
```diff
@@ -133,11 +133,15 @@
+        "START_WITH_SYSTEM_LABEL" : "Start WowUp with system",
+        "START_WITH_SYSTEM_DESCRIPTION" : "WowUp will start immediately after your operational ssytem is loaded",
+        "START_MINIMIZED_LABEL" : "Start WowUp minimized",
+        "START_MINIMIZED_DESCRIPTION" : "WowUp will start minimized on system startup"
```
Labels and descriptions for new options

linaori replied:
You can sync translations by running `npm run i18n`, this will add the missing translations to all other available translations, and sort them

Noxis-source commented on wowup-electron/src/assets/i18n/ru.json:
```diff
@@ -133,11 +133,15 @@
+        "START_WITH_SYSTEM_LABEL" : "Запускать WowUp вместе с ОС",
+        "START_WITH_SYSTEM_DESCRIPTION" : "WowUp будет запущен сразу после загрузки операционной системы",
+        "START_MINIMIZED_LABEL" : "Запускать WowUp свернутым",
+        "START_MINIMIZED_DESCRIPTION" : "WowUp будет запущен в свернутом виде со стартом системы"
```
Russian localization

Noxis-source commented on wowup-electron/src/common/constants.ts:
```diff
@@ -25,7 +25,7 @@ export const LAST_SELECTED_WOW_CLIENT_TYPE_PREFERENCE_KEY =
-
-// ERRORS
+export const START_WITH_SYSTEM_PREFERENCE_KEY = "start_with_system";
+export const START_MINIMIZED_PREFERENCE_KEY = "start_minimized"
```
Constants for new options

Noxis-source reviewed (commented):
Some notes

linaori reviewed (commented):
One minor thing that I think is missing; `setDefaultPreferences` sets the default settings in case nothing has been set yet (first run with this version). I think the default settings should be added there to avoid any weird behavior

Noxis-source commented:
@linaori 

> One minor thing that I think is missing; `setDefaultPreferences` sets the default settings in case nothing has been set yet (first run with this version). I think the default settings should be added there to avoid any weird behavior

I saw this part, but I thought that it makes sense to only set preferences that default to true? Right now settings that default to false are not in there.

>You can sync translations by running npm run i18n, this will add the missing translations to all other available translations, and sort them

Done.

jliddev commented on wowup-electron/main.ts:
```diff
@@ -230,7 +230,9 @@ function createWindow(): BrowserWindow {
 
   win.once("ready-to-show", () => {
-    win.show();
+    var startMinimized = (process.argv || []).indexOf('--hidden') !== -1;
```
Can we use something like minimist or yargs for argument parsing instead of hand making this?

Noxis-source replied:
Ok, will do.

P.S.
Why do you hate manual args parsing? :D

jliddev replied:
using a lib to do it will make it easier to add new args as we need and not have to overthink how to add them.

jliddev commented:
Have you investigated using the build in api to launch on start instead of auto-launch?
https://www.electronjs.org/docs/api/app#appsetloginitemsettingssettings-macos-windows

auto-launch looks to be abandoned maybe.

Noxis-source commented:
> 
> 
> Have you investigated using the built in api to launch on start instead of auto-launch?
> https://www.electronjs.org/docs/api/app#appsetloginitemsettingssettings-macos-windows
> 
> auto-launch looks to be abandoned maybe.

Well, yes - see the third comment =)
https://github.com/jliddev/WowUp/pull/219#issuecomment-715903191

In short, it does not support Linux at all. And besides auto-launch received its last commit in 2017, I do not think something might go wrong there, unless target platforms have big updates for their autostartup logics (whic they did not have, as far as I know). 

Another possible way is to do all this manually, but I did not want to rely on that because my knowledge of MacOs and Linux are fairly limited.

jliddev commented:
> > Have you investigated using the built in api to launch on start instead of auto-launch?
> > https://www.electronjs.org/docs/api/app#appsetloginitemsettingssettings-macos-windows
> > auto-launch looks to be abandoned maybe.
> 
> Well, yes - see the third comment =)
> [#219 (comment)](https://github.com/jliddev/WowUp/pull/219#issuecomment-715903191)
> 
> In short, it does not support Linux at all. And besides auto-launch received its last commit in 2017, I do not think something might go wrong there, unless target platforms have big updates for their autostartup logics (whic they did not have, as far as I know).
> 
> Another possible way is to do all this manually, but I did not want to rely on that because my knowledge of MacOs and Linux is fairly limited.

Fair point. Looking at the code for auto-launch the linux setup is just creating a text file. So my opinion would be to use the supported method via electron for windows and mac and if we're on linux use either just the same method as below or this package.
I would hesitate to use the lib when we have it built into electron since mac is constantly changing security stuff we should let them support it. auto-launch also has a bunch of open tickets from years ago of various issues.

https://github.com/Teamwork/node-auto-launch/blob/master/src/AutoLaunchLinux.coffee

Noxis-source commented:
Ok, done

Noxis-source commented:
I added data bindings and disable the second toggle based on the state of the first one. I also added value reset for this scenario. Corrected labels to be more clear.

jliddev commented on wowup-electron/src/app/services/wowup/wowup.service.ts:
```diff
@@ -27,13 +27,19 @@ import {
 import { UpdateCheckResult } from "electron-updater";
 
 const LATEST_VERSION_CACHE_KEY = "latest-version-response";
+const isMac = process.platform === "darwin";
```
Please use ElectronService instead.

jliddev commented on wowup-electron/src/app/services/wowup/wowup.service.ts:
```diff
@@ -265,4 +301,25 @@ export class WowUpService {
+        autoLauncher.disable();
+    }
+    else {
+      remote.app.setLoginItemSettings({
```
please use ElectronService

jliddev commented:
Looks good, just some minor tweaks for electron service stuff.

Noxis-source commented:
Done
